### PR TITLE
[Host] Refactor 'Retry' delegate in IConsumerErrorHandler as response type

### PR DESCRIPTION
**Pull request by EtherZa**
While the discussed 'pipeline reimagination' was a little too ambitious for the time that I have available; I have refactored the implementation to extract the `retry` delegate from `IConsumerErrorHandler.OnHandleError` and to rather include it as a response value. This does introduce a breaking change, but allows the handler to perform all of the previous logic albeit with a small change for retry.

The `OnHandlerError` method signature therefore changes from:
```cs
Task<ConsumerErrorHandlerResult> OnHandleError(T message, Func<Task<object>> retry, IConsumerContext consumerContext, Exception exception);
```
to:
```cs
// where `attempts` is the execution count
Task<ConsumerErrorHandlerResult> OnHandleError(T message, IConsumerContext consumerContext, Exception exception, int attempts);
```
In order to retry the execution (with a new scope as requried), the implemented method must return `ConsumerErrorHandlerResult.Retry`. Further to this, abstract implmentations of `IConsumerErrorHandler` have been added with methods to support `Failure()`, `Success(object? response = null)` and `Retry()` for a cleaner experience (similar to `return Ok();` with asp.net controllers).

Example:
```cs
public class RetryHandler<T> : ConsumerErrorHandler<T>
{
    private static readonly Random _random = new();

    public override async Task<ConsumerErrorHandlerResult> OnHandleError(T message, IConsumerContext consumerContext, Exception exception, int attempts)
    {
        if (attempts <= 3)
        {
            var delay = (attempts * 1000) + (_random.Next(1000) - 500);

            await Task.Delay(delay, consumerContext.CancellationToken);
            return Retry();
        }

        return Failure();
    }
}
```

### CODE REVIEW

File: docs/intro.t.md [hunk 1]
```diff
@@ -1059,14 +1059,17 @@ public class LoggingConsumerInterceptor<TMessage> : IConsumerInterceptor<TMessag
...
-> The `retry()` parameter allows the message processing pipeline, including consumer interceptors, to retry processing when transient errors occur and retries are desired.
+The returned `ConsumerErrorHandlerResult` object is used to override the execution for the remainder of the execution pipeline.
+| Result  | Description |
+|---------|-------------|
+| Failure | The message failed to be processed and should be returned to the queue |
+| Success | The pipeline must treat the message as having been processed successfully |
+| SuccessWithResponse | The pipeline to treat the messagage as having been processed successfully, returning the response to the request/response invocation ([IRequestResponseBus<T>](../src/SlimMessageBus/RequestResponse/IRequestResponseBus.cs)) |
+| Retry   | Execute the pipeline again (any delay/jitter should be applied before returning from method) |
 
 To enable SMB to recognize the error handler, it must be registered within the Microsoft Dependency Injection (MSDI) framework:
```
Comment (zarusz) [inline docs/intro.t.md:1075]: @EtherZa  could you also add the sample error handler from the PR into the into.t.md?
I think people will find it handy to have a starting sample.

Comment (zarusz) [reply docs/intro.t.md:1075]: Also please make a note about the fact the message scope will be re-created with each retry. Its not apparent when reading this description.

**Comment by zarusz** [comment]
This is an excellent and pragmatic approach to improving the error-handling pipeline while achieving the message scope recreation as discussed in issue #347. The implementation is mostly backward-compatible and could potentially outperform the `Func<Task> next` delegate approach.

I've suggested a few minor adjustments to the documentation for clarity and completeness.

Once those are addressed, I’ll be happy to approve the PR. Great work!

File: docs/intro.t.md (hunk 1 above)
Comment (EtherZa) [reply docs/intro.t.md:1077]: Sample and note added.

**Comment by zarusz** [comment]
It's been released under v3 rc:
https://www.nuget.org/packages/SlimMessageBus.Host.AzureServiceBus/3.0.0-rc901

**Pull request merged**